sdk/tests/conformance2/buffers/getBufferSubDataAsync.html

```diff
@@ -0,0 +1,249 @@
+** the following conditions:
+**
+** The above copyright notice and this permission notice shall be included
+** in all copies or substantial portions of the Materials.
+**
+** THE MATERIALS ARE PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+** EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+** MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+** IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+** CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+** TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+** MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+*/
+
+-->
+
+<!DOCTYPE html>
+<html>
+<head>
+<meta charset="utf-8">
+<title>WebGL getBufferSubDataAsync test.</title>
+<link rel="stylesheet" href="../../resources/js-test-style.css"/>
+<script src="../../js/js-test-pre.js"></script>
+<script src="../../js/webgl-test-utils.js"></script>
+</head>
+<body>
+<div id="description"></div>
+<div id="console"></div>
+<script>
+"use strict";
+description("This test makes sure that getBufferSubDataAsync acts as expected governed by WebGL 2.");
+
+var wtu = WebGLTestUtils;
+
+var gl = wtu.create3DContext(undefined, undefined, 2);
+
+var vertices = [
+     1.1,  1.0,  1.3,
+    -1.0, -1.0,  -5.0,
+     5.3, -1.0,  1.0
+];
+var floatArray = new Float32Array(vertices);
+var dstData = new Float32Array(vertices.length);
+var extraLargeBuffer = new Float32Array(vertices.length + 1);
+var resolvedArray;
+var promise;
+
+function clearDstData() {
+    dstData = new Float32Array(vertices.length);
+}
+
+wtu.glErrorShouldBe(gl, gl.NO_ERROR, "Should be no errors from setup.");
+
+function expectRejected(promise, message) {
+  return promise.then(function(v) {
+    testFailed("should be rejected but was " + v + ": " + message);
+  }, function(e) {
+    testPassed("correctly rejected with " + e + ": " + message);
+  });
+}
+
+var types = [
+  'ARRAY_BUFFER',
+  'ELEMENT_ARRAY_BUFFER',
+  'COPY_READ_BUFFER',
+  'COPY_WRITE_BUFFER',
+  'PIXEL_PACK_BUFFER',
+  'PIXEL_UNPACK_BUFFER',
+  'TRANSFORM_FEEDBACK_BUFFER',
+  'UNIFORM_BUFFER'
+];
+var buffers = [];
+
+var buffer;
+for (var i = 0; i < types.length; i++) {
+  buffer = gl.createBuffer();
+  gl.bindBuffer(gl[types[i]], buffer);
+  gl.bufferData(gl[types[i]], floatArray, gl.STATIC_DRAW);
+  buffers.push(buffer);
+}
+
+// This long promise chain is to serialize the tests - so that the errors
+// associated with a given test always appear underneath that test.
+// For each test in this chain, if it performs test checks in a promise,
+// it MUST return that promise so that it completes before the next test.
+var allPromises = Promise.resolve();
+
+allPromises = allPromises.then(function() {
+  debug("Testing promise rejection");
+});
+
+allPromises = allPromises.then(function() {
+  var code = "promise = gl.getBufferSubDataAsync(gl.ARRAY_BUFFER, 0, new ArrayBuffer(4))";
```

Comment:
Those are type checks, which I didn't make clear at all. I added debug prints to explain.

```
PASS correctly rejected with TypeError: Failed to execute 'getBufferSubDataAsync' on 'WebGL2RenderingContext': parameter 3 is not of type 'ArrayBufferView'.: promise = gl.getBufferSubDataAsync(gl.ARRAY_BUFFER, 0, new ArrayBuffer(4))
```

```
PASS correctly rejected with TypeError: Failed to execute 'getBufferSubDataAsync' on 'WebGL2RenderingContext': parameter 3 is not of type 'ArrayBufferView'.: promise = gl.getBufferSubDataAsync(gl.ARRAY_BUFFER, 0, null)
```